runtime/near-vm-runner/src/imports.rs

```diff
@@ -323,8 +323,9 @@ wrapped_imports! {
     storage_iter_prefix<[prefix_len: u64, prefix_ptr: u64] -> [u64]>,
     storage_iter_range<[start_len: u64, start_ptr: u64, end_len: u64, end_ptr: u64] -> [u64]>,
     storage_iter_next<[iterator_id: u64, key_register_id: u64, value_register_id: u64] -> [u64]>,
-    // Function for the injected gas counter. Automatically called by the gas meter.
+    // Function for the injected gas counter. Automatically called by the gas meter.\
     gas<[gas_amount: u32] -> []>,
+    out_of_gas_callback<[] -> []>,
```

Comment:
I'll do some research on how to prevent this function been called, possibly a pass before inject gas, statically check no out_of_gas_callback should be called. There was no good reason for user to call gas, because it simply spend you more gas but do nothing, and you're right it's less harm. Out_of_gas_callback if called manually would cause GasExceed but actually not exceed.

Reply:
One idea (not sure if good or not) is to take advantage of the fact that wasm imports are two-level. We can use `env.foo` for host functions, and `private.foo` for "our stuff".

Reply:
@matklad I started to address this problem. So import it in `private` env should work, because we have a preparation pass to check all imports are from env pass. We can, doing this pass before inject_gas_counter so this prevent user code from calling `private.out_of_gas_callback`. Then inject gas counter, which will add some `call private.out_of_gas_callback` instruction, which is the correct behavior.

However `out_of_gas_callback` implementation has been modified. it's now:
```
    pub fn out_of_gas_callback(&mut self, last_ops: u64) -> Result<()> {
        let value = last_ops * Gas::from(self.config.regular_op_cost);
        self.gas_counter.pay_wasm_gas(value)
    }
```

Which is almost identical to `env.gas`:
```
    pub fn gas(&mut self, gas_amount: u32) -> Result<()> {
        let value = Gas::from(gas_amount) * Gas::from(self.config.regular_op_cost);
        self.gas_counter.pay_wasm_gas(value)
    }
```

So based on your point, I believe it is now bening to leave out_of_gas_callback in env. WDYT?


(The reason to change is implementation is to pass the last wasm gas (ops) that would cause underflow in wasm gas meter back to host, so that host can correctly figure out final gas usage in GasExceeded case) If user try to call out_of_gas_callback manually, it will just waste some gas and not necessarily run out of gas, same as `gas`.

Reply:
>WDYT?

Sounds good to me! In any case, even if we do decide to do some extra future-proofing, that can be tackled in a separate PR.

Looking at the above, one new (probably bad) idea I have is that, if `gas` and `out_of_gas_callback` have identical signaures, we might be able to use just a single import for both? That is, we only call one or the other, so we can add a flag to VM Logic which selects the meaning of `gas` fn.

Reply:
Right, they are a bit difference in signature (u32 vs u64). This is to adapt the old gas_counter and new gas_couner without a conversion inst. (old use wasm i32, is sufficient to count used. and new gas_counter counts remaining, wasm i64 is required to avoid overflow in some test cases)

We will be able to just import a single of them, depending on the protocol version only one of them will be needed